README.md
# bcbsn

[![Build Status](https://travis-ci.org/030/bcbsn.svg?branch=master)](https://travis-ci.org/030/bcbsn)
[![Go Report Card](https://goreportcard.com/badge/github.com/030/bcbsn)](https://goreportcard.com/report/github.com/030/bcbsn)
![DevOps SE Questions](https://img.shields.io/stackexchange/devops/t/bcbsn.svg)
![Docker Pulls](https://img.shields.io/docker/pulls/utrecht/bcbsn.svg)
![Issues](https://img.shields.io/github/issues-raw/030/bcbsn.svg)
![Pull requests](https://img.shields.io/github/issues-pr-raw/030/bcbsn.svg)
![Total downloads](https://img.shields.io/github/downloads/030/bcbsn/total.svg)
![License](https://img.shields.io/github/license/030/bcbsn.svg)
![Repository Size](https://img.shields.io/github/repo-size/030/bcbsn.svg)
![Contributors](https://img.shields.io/github/contributors/030/bcbsn.svg)
![Commit activity](https://img.shields.io/github/commit-activity/m/030/bcbsn.svg)
![Last commit](https://img.shields.io/github/last-commit/030/bcbsn.svg)
![Release date](https://img.shields.io/github/release-date/030/bcbsn.svg)
![Latest Production Release Version](https://img.shields.io/github/release/030/bcbsn.svg)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=bugs)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=code_smells)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=coverage)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Duplicated Lines (%)](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=duplicated_lines_density)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=ncloc)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=sqale_rating)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=alert_status)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Reliability Rating](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=reliability_rating)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=security_rating)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=sqale_index)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=030_bcbsn&metric=vulnerabilities)](https://sonarcloud.io/dashboard?id=030_bcbsn)
[![CII Best Practices](https://bestpractices.coreinfrastructure.org/projects/2809/badge)](https://bestpractices.coreinfrastructure.org/projects/2809)

The Bitbucket Cloud Build Status Notifier (BCBSN) is a tool that sends a build status to bitbucket cloud.
Oauth credentials are required to perform this action and these have to be created in the user settings
menu that resides in bitbucket cloud. Note: a Callback URL, e.g. https://ci-system-name is required in order
to create Oauth credentials.

In order to run this tool, one could choose to download the binary from the releases section in this
repository or use docker.

Example for running this command given:

- **client id**: 1234
- **client secret**: 4567
- **build state**: INPROGRESS
- **git commit hash**: f76d123498a053c1789057a41d6c3fcvg8b49cd7
- **owner**: 030
- **repository name**: bcbsn
- **build number**: 70
- **build url**: https://travis-ci.org/030/bcbsn/builds/523263434

[![dockeri.co](https://dockeri.co/image/utrecht/bcbsn)](https://hub.docker.com/r/utrecht/bcbsn)

To print the help message

```
docker run utrecht/bcbsn:2.0.2 -h
```

returns:

```
Usage of /usr/local/bcbsn:
  -clientID string
    	The clientID used for the 'client credentials' token flow with BitBucket
  -clientSecret string
    	The clientSecret used for the 'client credentials' token flow with BitBucket
  -commit string
    	The commit, e.g. 57484fd5460017aef111e8b4ec116a30ff0b4904
  -key string
    	The key, e.g. a unique id of the build (use the build id)
  -name string
    	An identifier for the build e.g. 'build 2'
  -owner string
    	The owner of the repository, e.g. it is 'atlassian' in 'https://bitbucket.org/atlassian/stash-example-plugin/src/master/'
  -repoSlug string
    	The repoSlug, e.g. some-repository
  -state string
    	The state, e.g. SUCCESSFUL, INPROGRESS or FAILED
  -url string
    	The url, e.g. https://travis-ci.org/030/bcbsn/builds/523263434
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.2] - 2019-05-21
### Fixed
- Accept either a 200 or 201 as a HTTP response status code [@jorianvo]

## [2.0.1] - 2019-05-14
### Fixed
- List of arguments required to use this tool updated.
- Publication of artifacts.

## [2.0.0] - 2019-05-12
### Added
- HTTP tests [@jorianvo](https://github.com/jorianvo).
- Multiple badges including devops.stackexchange batch about [bcbsn tag](https://devops.stackexchange.com/questions/tagged/bcbsn).
- New dockerhub repository [utrecht/bcbsn](https://cloud.docker.com/u/utrecht/repository/docker/utrecht/bcbsn) as it was
impossible to rename repositories on dockerhub.

### Changed
- Repository renamed to bcbsn as former name exceeded tag character limit on [StackExchange](https://stackexchange.com/).
- All references in README renamed due to repository name change.

## [1.1.2] - 2019-05-06
### Added
- Documented how to use the bcbsn tool.

## [1.1.1] - 2019-04-23
### Fixed
- Executable file in Docker image was incorrect.

## [1.1.0] - 2019-04-23
### Changed
- Improved logging to ensure that it will become easier to use the bcbsn tool.

## [1.0.2] - 2019-03-26
### Fixed
- Access token missing due to missing CA certificates in Scratch image.

## [1.0.1] - 2019-03-17
### Changed
- Use Scratch as a base to get a smaller docker image.

## [1.0.0] - 2019-03-17
### Added
- Send a build status to bitbucket cloud.

[Unreleased]: https://github.com/030/bcbsn/compare/2.0.2...HEAD
[2.0.2]: https://github.com/030/bcbsn/compare/2.0.1...2.0.2
[2.0.1]: https://github.com/030/bcbsn/compare/2.0.0...2.0.1
[2.0.0]: https://github.com/030/bcbsn/compare/1.1.2...2.0.0
[1.1.2]: https://github.com/030/bcbsn/compare/1.1.1...1.1.2
[1.1.1]: https://github.com/030/bcbsn/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/030/bcbsn/compare/1.0.2...1.1.0
[1.0.2]: https://github.com/030/bcbsn/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/030/bcbsn/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/030/bcbsn/releases/tag/1.0.0


main_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"reflect"
	"strings"
	"testing"
	"time"
)

const (
	clientID     = "1234"
	clientSecret = "5678"
	commit       = "c1o2m3m4i5t6"
	key          = "1"
	url          = "http://build.url"
	state        = "SUCCESSFUL"
	name         = "build #1"
	owner        = "030"
	repoSlug     = "repo_slug"
)

var testHTTPClient = http.Client{
	Timeout: time.Second,
}

func TestHTTPClientShouldTimeOutAfterTenSeconds(t *testing.T) {
	expectedError := "net/http: request canceled (Client.Timeout exceeded while awaiting headers)"
	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Second * 2)
	}))
	defer svr.Close()

	err := setBuildStatusImpl(testHTTPClient, svr.URL, owner, repoSlug, commit, key, url, state, name)

	if !strings.Contains(err.Error(), expectedError) {
		t.Errorf("Expected '%v', but got '%v'", expectedError, err)
	}
}

func TestHTTPClientShouldFailIfResponseIsNotA201(t *testing.T) {
	expectedError := "Expected 200 or 201, but got 400 Bad Request"
	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
	}))
	defer svr.Close()

	err := setBuildStatusImpl(testHTTPClient, svr.URL, owner, repoSlug, commit, key, url, state, name)

	if err.Error() != expectedError {
		t.Errorf("Expected '%v', but got '%v'", expectedError, err)
	}
}

func TestHTTPClientShouldAcceptAnHTTP200(t *testing.T) {
	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	}))
	defer svr.Close()

	err := setBuildStatusImpl(testHTTPClient, svr.URL, owner, repoSlug, commit, key, url, state, name)
	if err != nil {
		t.Errorf("Expected no error, but got '%v'", err)
	}
}

func TestHTTPClientShouldAcceptAnHTTP201(t *testing.T) {
	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusCreated)
	}))
	defer svr.Close()

	err := setBuildStatusImpl(testHTTPClient, svr.URL, owner, repoSlug, commit, key, url, state, name)
	if err != nil {
		t.Errorf("Expected no error, but got '%v'", err)
	}
}

func TestPostBodyHasExpectedContent(t *testing.T) {
	expectedURL := "/030/repo_slug/commit/c1o2m3m4i5t6/statuses/build"
	expectedBody := Body{commit, key, url, state, name}
	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)

		if r.Method != "POST" {
			t.Errorf("Expected ‘POST’ request, got ‘%s’", r.Method)
		}
		if r.URL.EscapedPath() != expectedURL {
			t.Errorf("Expected request to '%s', got '%s'", expectedURL, r.URL.EscapedPath())
		}
		if r.Header.Get("Content-type") != "application/json" {
			t.Errorf("Expected content-type to be 'application/json', but got '%s'", r.Header.Get("Content-type"))
		}

		decoder := json.NewDecoder(r.Body)
		var actualBody Body
		err := decoder.Decode(&actualBody)
		if err != nil {
			t.Errorf("Error while reading request JSON: %s", err)
		}

		if !reflect.DeepEqual(expectedBody, actualBody) {
			t.Errorf("Expected request body '%s', but was '%s'", expectedBody, actualBody)
		}
	}))
	defer svr.Close()

	setBuildStatusImpl(testHTTPClient, svr.URL, owner, repoSlug, commit, key, url, state, name)
}


main.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"

	"golang.org/x/oauth2"
	"golang.org/x/oauth2/clientcredentials"
)

const bitbucketAPIEndpoint = "https://api.bitbucket.org/2.0/repositories"

type Body struct {
	UUID  string `json:"uuid"`  // The Git commit hash
	Key   string `json:"key"`   // The Build ID
	URL   string `json:"url"`   // The URL of the specific build
	State string `json:"state"` // The state of the build
	Name  string `json:"name"`  // The identifier of the build
}

var httpClientWithTimeout = http.Client{
	Timeout: time.Second * 10,
}

func prepareBody(commit, key, url, state, name string) ([]byte, error) {
	b := Body{commit, key, url, state, name}
	return json.Marshal(b)
}

func prepareOAUTH2HTTPClient(httpClient http.Client, clientID, clientSecret string) *http.Client {
	config := &clientcredentials.Config{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		TokenURL:     "https://bitbucket.org/site/oauth2/access_token",
	}

	// Instead of the default httpClient use our custom client (which sets a timeout)
	ctx := context.WithValue(oauth2.NoContext, oauth2.HTTPClient, httpClient)
	return config.Client(ctx)
}

// We need to be able to overwrite the endpoint in our unit tests
// So we use the approach as described here: https://stackoverflow.com/a/33774754/10224882
func setBuildStatus(clientID, clientSecret, owner, repoSlug, commit, key, url, state, name string) error {
	httpClient := prepareOAUTH2HTTPClient(httpClientWithTimeout, clientID, clientSecret)
	return setBuildStatusImpl(*httpClient, bitbucketAPIEndpoint, owner, repoSlug, commit, key, url, state, name)
}

func setBuildStatusImpl(httpClient http.Client, bitbucketEndpoint, owner, repoSlug, commit, key, url, state, name string) error {
	body, err := prepareBody(commit, key, url, state, name)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", bitbucketEndpoint+"/"+owner+"/"+repoSlug+"/commit/"+commit+"/statuses/build", bytes.NewBuffer(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := httpClient.Do(req)
	if resp != nil {
		// Bitbucket returns a 201 when it first creates the build badge all subsequent
		// requests (i.e. updates on the same build) return a 200. We can't write
		// it as r != b || r != c as go vet rejects this, see https://github.com/golang/go/issues/28446
		if !(resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusCreated) {
			return fmt.Errorf("Expected 200 or 201, but got %s", resp.Status)
		}
		defer resp.Body.Close()
	}

	if err != nil {
		return err
	}

	bodyBytes, _ := ioutil.ReadAll(resp.Body)
	fmt.Println(string(bodyBytes))

	return nil
}

func main() {
	var clientID = flag.String(
		"clientID",
		"",
		"The clientID used for the 'client credentials' token flow with BitBucket")

	var clientSecret = flag.String(
		"clientSecret",
		"",
		"The clientSecret used for the 'client credentials' token flow with BitBucket")

	var state = flag.String(
		"state",
		"",
		"The state, e.g. SUCCESSFUL, INPROGRESS or FAILED")

	var commit = flag.String(
		"commit",
		"",
		"The commit, e.g. 57484fd5460017aef111e8b4ec116a30ff0b4904")

	var owner = flag.String(
		"owner",
		"",
		"The owner of the repository, e.g. it is 'atlassian' in 'https://bitbucket.org/atlassian/stash-example-plugin/src/master/'")

	var repoSlug = flag.String(
		"repoSlug",
		"",
		"The repoSlug, e.g. some-repository")

	var key = flag.String(
		"key",
		"",
		"The key, e.g. a unique id of the build (use the build id)")

	var url = flag.String(
		"url",
		"",
		"The url, e.g. https://travis-ci.org/030/bcbsn/builds/523263434")

	var name = flag.String(
		"name",
		"",
		"An identifier for the build e.g. 'build 2'")

	flag.Parse()

	if *clientID == "" || *clientSecret == "" || *state == "" || *commit == "" || *owner == "" || *repoSlug == "" || *key == "" || *url == "" || *name == "" {
		flag.Usage()
		os.Exit(1)
	}

	err := setBuildStatus(*clientID, *clientSecret, *owner, *repoSlug, *commit, *key, *url, *state, *name)
	if err != nil {
		log.Fatal(err)
	}
}
